### ISSUE TITLE
Immutable JS?

### ISSUE THREAD

**Issue by skellock** [infinitered/reactotron#242]
Although I don't personally use ImmutableJS, I know people do.

Rather than start a campaign of terror to get them to switch to `seamless-immutable`, figure out what it would take to support it.

@andrewvy [championed this a while ago](https://github.com/reactotron/reactotron/pull/21) but I ninja-refactored the code before it was committed.

There have been a few people asking about it.

I feel like there should be a "serializer" somewhere in the ImmutableJS community that does this?  I just don't know of one.  Maybe it's out of the box?

Anyone have any insight into this?

**Comment by jdmunro** [comment]
What do you need to serialise to/from? By default you can serialise to and from string, but you'll lose _some_ typing information if you use special immutable types such as Records, Sets. By default you'll get everything back as Maps and Lists.

Though I've never used it, there's this serialiser which supposedly preserves the extra type information: https://github.com/glenjamin/transit-immutable-js

**Comment by skellock** [comment]
Maybe there isn't a problem then?  Maybe I'm just asking the wrong question.  Perhaps this is a querying question.

If I can't dot my way through a immutable object like `cart.items.7.price`, how would I traverse their structures to get at that answer?  I'd just have to read into that I guess 😀

**Comment by jdmunro** [comment]
There's a nice way to access key paths of an immutable object :)

You can do this:

```
myObject.getIn(['cart', 'items' , 7, 'price'])
```

So in theory you'd just need to convert a string query into an array key path as above. For that bit of functionality, anyway!

**Comment by jdmunro** [comment]
I don't have much time to contribute to development but I'd be happy to help out with testing any WIP implementations of support for ImmutableJS 👍

**Comment by skellock** [comment]
`getIn` is exactly what I'm looking for.  Thank you!

**Comment by jsvg** [comment]
has #315 covered this?

**Comment by skellock** [comment]
Not yet.  Thanks for the reminder.  We've got another release happening shortly.  I'll see to it that @rmevans9 's work implementing it sees it in!

**Comment by jsvg** [comment]
I think there might still be a small problem with the implementation. It looks like errors are still thrown when passing actions with immutable payloads. I don't have too much time to dig into a proper solution, but dropping this snippet in `report-action.js` before line 11 (`reactotron.send`) seems to work in my case.

```js
for (var key in action) {
    if (!action.hasOwnProperty(key)) continue;
    if (action[key] && action[key].toJS) {
    action[key] = action[key].toJS();
      continue;
    }
  }
```

Hope this helps!

**Comment by rmevans9** [comment]
Yeah, the current implementation only considers the redux state tree as immutable, should be an easy fix to make it work for actions too

**Comment by jsvg** [comment]
Just noticed that the code above mutates actions so the cleaned actions should be `Object.assign`ed to a new var.

```js
const newAction = Object.assign({}, action);
for (const key in action) {
  if (!action.hasOwnProperty(key)) continue;
  if (action[key] && action[key].toJS) {
  newAction[key] = action[key].toJS();
    continue;
  }
}
```

**Comment by skellock** [comment]
This is fixed in 1.9.  Sorry about that.
